### PR TITLE
Add tools.cppstd_minimum_required

### PR DESCRIPTION
**Pull request by uilianries**
Related PR: https://github.com/conan-io/conan/pull/5997

### CODE REVIEW

File: reference/tools.rst [hunk 1]
```diff
@@ -1622,3 +1622,73 @@ Converts Conan style architecture into Android NDK style architecture.
 
 Parameters:
     - **arch** (Required): Arch to perform the conversion. Usually this would be ``self.settings.arch``.
+
+.. _tools.check_min_cppstd:
+
+tools.check_min_cppstd()
+-------------------------------
```
Comment (lasote) [inline reference/tools.rst]: too long

File: reference/tools.rst [hunk 2]
```diff
@@ -1622,3 +1622,73 @@ Converts Conan style architecture into Android NDK style architecture.
...
+
+.. code-block:: python
+
+    def check_min_cppstd(conanfile, cppstd, gnu_extensions=False)
+
+Validate the current cppstd from settings or compiler, if it is supported by the required cppstd version.
+It raises a ``InvalidConfiguration`` when is not supported.
```
Comment (lasote) [inline reference/tools.rst]: That exception doesn't exist. Is ``ConanInvalidConfiguration``. Link to https://docs.conan.io/en/latest/mastering/conditional.html

File: reference/tools.rst [hunk 3]
```diff
@@ -1622,3 +1622,73 @@ Converts Conan style architecture into Android NDK style architecture.
...
+
+    from conans import tools, ConanFile
+
+    class Recipe(ConanFile):
+        ...
+
+        def configure(self):
+            tools.check_min_cppstd(self, "17")
+
+* If the current cppstd does not support C++17, ``check_min_cppstd`` will raise an ``InvalidConfiguration`` error.
```
Comment (lasote) [inline reference/tools.rst]: Idem

File: reference/tools.rst [hunk 4]
```diff
@@ -1622,3 +1622,73 @@ Converts Conan style architecture into Android NDK style architecture.
...
+* When there is no cppstd declared or is ``None``, the current compiler version listed in the profile will be used to detect
+  the compatibility with the required C++ standard.
+* If ``gnu_extensions`` is True and your current OS is Linux, the settings and/or compiler must support/offer GNU extensions
+  (e.g. gnu17), otherwise, an ``InvalidConfiguration`` will be raised. The ``gnu_extensions`` is only checked on Linux, for
```
Comment (lasote) [inline reference/tools.rst]: Idem

File: reference/tools.rst [hunk 5]
```diff
@@ -1622,3 +1622,73 @@ Converts Conan style architecture into Android NDK style architecture.
...
+
+Parameters:
+    - **conanfile** (Required): ConanFile instance. Usually ``self``.
+    - **cppstd** (Required): C++ standard version which must be supported.
+    - **gnu_extensions** (Optional): GNU extension is required (only for Linux).
+
+.. _tools.valid_min_cppstd:
+
+tools.valid_min_cppstd()
+-------------------------------
```
Comment (lasote) [inline reference/tools.rst]: too long

File: reference/tools.rst [hunk 6]
```diff
@@ -1622,3 +1622,73 @@ Converts Conan style architecture into Android NDK style architecture.
...
+    from conans import tools, ConanFile
+
+    class Recipe(ConanFile):
+        ...
+
+        def configure(self):
+            if not tools.valid_min_cppstd(self, "17"):
+                self.output.error("C++17 is required.")
+
+* The ``valid_min_cppstd`` works exactly like ``check_min_cppstd``, however, it does not raise ``InvalidConfiguration`` error.
```
Comment (lasote) [inline reference/tools.rst]: idem

File: reference/tools.rst [hunk 7]
```diff
@@ -1622,3 +1622,73 @@ Converts Conan style architecture into Android NDK style architecture.
...
+* If the current cppstd does not support C++17, ``valid_min_cppstd`` returns ``False``.
+* When there is no cppstd declared or is ``None``, the current compiler version listed in the profile will be used to detect
```
Comment (lasote) [inline reference/tools.rst]: I would remove from here. Saying it is the same but returning False when the other raises, it is enough

Comment (uilianries) [reply reference/tools.rst]: done

File: reference/tools.rst (hunk 6 above)
Comment (uilianries) [reply reference/tools.rst]: done

File: reference/tools.rst (hunk 5 above)
Comment (uilianries) [reply reference/tools.rst]: done

Comment (uilianries) [reply reference/tools.rst]: done

File: reference/tools.rst (hunk 4 above)
Comment (uilianries) [reply reference/tools.rst]: done

File: reference/tools.rst (hunk 3 above)
Comment (uilianries) [reply reference/tools.rst]: done

File: reference/tools.rst (hunk 2 above)
Comment (uilianries) [reply reference/tools.rst]: done

File: reference/tools.rst (hunk 1 above)
Comment (uilianries) [reply reference/tools.rst]: done

File: reference/tools.rst [hunk 8]
```diff
@@ -1622,3 +1622,69 @@ Converts Conan style architecture into Android NDK style architecture.
...
+.. _tools.check_min_cppstd:
+
+tools.check_min_cppstd()
+------------------------
+
+.. code-block:: python
+
+    def check_min_cppstd(conanfile, cppstd, gnu_extensions=False)
+
+Validate the current cppstd from settings or compiler, if it is supported by the required cppstd version.
```
Comment (lasote) [inline reference/tools.rst]:
```suggestion
Validates if the applied cppstd setting (from `compiler.cppstd` settings or deducing the default from `compiler` and `compiler.version`) is at least the value specified in the `cppstd` argument.
```

File: reference/tools.rst [hunk 9]
```diff
@@ -1622,3 +1622,69 @@ Converts Conan style architecture into Android NDK style architecture.
...
+    from conans import tools, ConanFile
+
+    class Recipe(ConanFile):
+        ...
+
+        def configure(self):
+            tools.check_min_cppstd(self, "17")
+
+* If the current cppstd does not support C++17, ``check_min_cppstd`` will raise an ``ConanInvalidConfiguration`` error.
+* When there is no cppstd declared in setttings or is ``None``, the current compiler version listed in the profile will be used to detect
```
Comment (lasote) [inline reference/tools.rst]:
```suggestion

```

File: reference/tools.rst [hunk 10]
```diff
@@ -1622,3 +1622,69 @@ Converts Conan style architecture into Android NDK style architecture.
...
+  the default standard C++ version supported.
```
Comment (lasote) [inline reference/tools.rst]:
```suggestion

```

File: reference/tools.rst [hunk 11]
```diff
@@ -1622,3 +1622,69 @@ Converts Conan style architecture into Android NDK style architecture.
...
+* If ``gnu_extensions`` is True, the settings and/or compiler must support/offer GNU extensions
```
Comment (lasote) [inline reference/tools.rst]:
```suggestion
* If ``gnu_extensions`` is True, it is required that the applied ``cppstd`` supports the gnu extensions.
```

File: reference/tools.rst [hunk 12]
```diff
@@ -1622,3 +1622,69 @@ Converts Conan style architecture into Android NDK style architecture.
...
+
+    class Recipe(ConanFile):
+        ...
+
+        def configure(self):
+            if not tools.valid_min_cppstd(self, "17"):
+                self.output.error("C++17 is required.")
+
+* The ``valid_min_cppstd`` works exactly like ``check_min_cppstd``, however, it does not raise ``ConanInvalidConfiguration`` error.
+* If the current cppstd does not support C++17, ``valid_min_cppstd`` returns ``False``.
```
Comment (lasote) [inline reference/tools.rst]:
```suggestion

```

File: reference/tools.rst [hunk 13]
```diff
@@ -1622,3 +1622,69 @@ Converts Conan style architecture into Android NDK style architecture.
...
+* If ``gnu_extensions`` is True, the settings and/or compiler must support/offer GNU extensions
```
Comment (lasote) [inline reference/tools.rst]:
```suggestion

```

File: reference/tools.rst [hunk 14]
```diff
@@ -1622,3 +1622,69 @@ Converts Conan style architecture into Android NDK style architecture.
...
+  (e.g. gnu17), otherwise, it will return ``False``. The ``gnu_extensions`` is checked in any OS.
```
Comment (lasote) [inline reference/tools.rst]:
```suggestion

```

**Pull request merged**